channeldb/db.go

```diff
@@ -263,3 +296,123 @@ func (d *DB) FetchAllChannels() ([]*OpenChannel, error) {
 
 	return channels, err
 }
+
+// SyncVersions function is used for safe db version synchronization. It applies
+// migration functions to the current database and recovers the previous
+// state of db if at least one error/panic appeared during migration.
+func (d *DB) SyncVersions(versions []Version) (returnErr error) {
+	meta, err := d.FetchMeta()
+	if err != nil {
+		return err
+	}
+
+	cleanup, err := d.createBackup()
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if r := recover(); r != nil {
+			returnErr = fmt.Errorf("panic occured: %v",
+				errors.New(r))
+		}
+
+		if returnErr != nil {
+			if err := cleanup(true); err != nil {
+				returnErr = fmt.Errorf("while restoring db "+
+					"because of error: '%v', another "+
+					"error is occured: '%v'", returnErr, err)
+			}
+		} else {
+			if err := cleanup(false); err != nil {
+				returnErr = fmt.Errorf("migration was proceeded "+
+					"without problems, but while backup "+
+					"cleanup error is occured: %v", err)
+			}
+		}
+	}()
+
+	latestVersion := getLatestDBVersion(versions)
+
+	if meta.dbVersionNumber < latestVersion {
```

Comment:
In order to make the migration process more robust and fault-tolerant, the entire migration process should take place within a single `bolt` database transaction.